quilt-tester/quilt-tester.go

```
@@ -0,0 +1,613 @@
+	}
+
+	l.logInfo("Finished Tests", infoMsg)
+	l.logInfo(fmt.Sprintf("Finished Test Suite: %s", ts.name), infoMsg)
+
+	// Kill the current quilt process
+	if err := cmd.Process.Kill(); err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+func runTest(test string, m map[string]string) error {
+	dir, file := filepath.Split(test)
+	contents, err := fileContents(filepath.Join(dir, "src", file, file) + ".go")
+	if err != nil {
+		l.logInfo(fmt.Sprintf("Could not read test %s", test), infoMsg)
+		l.logInfo(err.Error(), errMsg)
+		return err
+	}
+
+	testLog := file + "-" + m["publicIP"] + ".txt"
+	l.currTest = filepath.Join(l.passedDir, testLog)
+
+	scp(m["publicIP"], test, file)
+	sshCmd := sshGen(m["publicIP"], exec.Command(fmt.Sprintf("./%s", file)))
+	output, err := sshCmd.CombinedOutput()
+	if err != nil || !strings.Contains(string(output), "PASSED") {
+		l.currTest = filepath.Join(l.failedDir, testLog)
+		l.logTest("Failed!", infoMsg)
+		if err == nil {
+			err = errors.New("failed test")
+		}
+	}
+
+	l.logTest("Begin test source", infoMsg)
+	l.logTest(contents, verbMsg)
+	l.logTest("End test source", infoMsg)
+	l.logTest("Begin test output", infoMsg)
+	l.logTest(string(output), verbMsg)
+	fmt.Println(string(output))
+	l.logTest("End test output", infoMsg)
+	return err
+}
+
+func runSpec(spec string) (string, *exec.Cmd, error) {
+	cmd := exec.Command("/quilt", "run", spec)
+	if spec == configSpecName {
+		output, err := runAndWaitFor(cmd, "New connection.", 500)
+		return output, cmd, err
+	}
+	l.logContainer(fmt.Sprintf("Running %v for 60 seconds", cmd.Args), infoMsg)
+	cmd.Start()
+	time.Sleep(60 * time.Second)
+	return "", cmd, nil
+}
+
+func runAndWaitFor(cmd *exec.Cmd, trigger string, timeout int) (string, error) {
+	l.logContainer(fmt.Sprintf("Running %v until %s", cmd.Args, trigger), infoMsg)
+	output := ""
+	pipe, err := cmd.StderrPipe()
+	if err != nil {
+		panic(err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		panic(err)
+	}
+
+	done := make(chan error, 1)
```

What if the command finishes and send to `done`, then the goroutine eats the `done` signal? The select at the bottom will just be left waiting around for the timer to go off. Or worse, the select at the bottom eats the error from `cmd.Wait()` and the goroutine is left running potentially forever.